velox/functions/prestosql/aggregates/ReduceAgg.cpp

```diff
@@ -0,0 +1,537 @@
+//   Global aggregation would go slightly differently.
+//
+//   Consider dataset {1, 2, 3, 4, 5, 6,..100}.
+//
+//   This implementation goes like so:
+//
+//   s0 - initialValue, s - state of the only group.
+//
+//   s = s0
+//   s = f(s, 1)
+//   s = f(s, 2)
+//   s = f(s, 3)
+//   s = f(s, 4)
+//   s = f(s, 5)
+//   ...
+//
+//   The inputFunction lambda expression is evaluated 100 times, once per row.
+//
+//   A more efficient approach would be:
+//
+//   f - inputFunction, g - combineFunction
+//
+//   Convert all inputs into states:
+//
+//   [s1, s2, s3, s4, s5,..s100] = f([s0, s0, s0, s0, s0,..], [1, 2, 3, 4, 5,..100])
+//
+//   Combine 100 states into 50:
+//
+//   [s1, s2,...s50] = g([s1, s3,...s99], [s2, s4,...s100])
+//
+//   Combine these 50 states into 25 states:
+//
+//   [s1, s2,...s25] = g([s1, s3,...s49], [s2, s4,...s50])
+//
+//   Continue in this manner until all states are combined. This requires only
+//   log2(100), ~7, expression evaluations.
+//
+// Note that the more efficient algorithm for global aggregation doesn't support
+// applying reduce_agg to sorted inputs.
+//
+// Also, note that the more efficient algorithm for global aggregation can be
+// used in a group-by as well. Each evaluation of the lambda could be reducing
+// the cardinality of each group by a factor of 2. This would be require
+// log2(max-group-cardinality) evaluations.
+//
+// A common use case for reduce_agg is to compute a product of input values.
+//
+//   reduce_agg(x, 1, (a, b) -> (a * b), (a, b) -> (a * b))
+//
+// In this case, the best option is to identify this pattern and invoke
+// specialized 'product(x)' aggregate function instead.
+//
+// clang-format on
+class ReduceAgg : public exec::Aggregate {
+ public:
+  ReduceAgg(TypePtr resultType) : exec::Aggregate(std::move(resultType)) {}
+
+  bool isFixedSize() const override {
+    return false;
+  }
+
+  int32_t accumulatorFixedWidthSize() const override {
+    return sizeof(ReduceAggAccumulator);
+  }
+
+  void initializeNewGroups(
+      char** groups,
+      folly::Range<const vector_size_t*> indices) override {
+    for (auto i : indices) {
+      new (groups[i] + offset_) ReduceAggAccumulator();
+    }
+  }
+
+  void extractValues(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    VELOX_CHECK(result);
+    (*result)->resize(numGroups);
+
+    auto* rawNulls = exec::Aggregate::getRawNulls(result->get());
+
+    for (int32_t i = 0; i < numGroups; ++i) {
+      char* group = groups[i];
+      auto accumulator = value<ReduceAggAccumulator>(group);
+      if (!accumulator->hasValue()) {
+        (*result)->setNull(i, true);
+      } else {
+        exec::Aggregate::clearNull(rawNulls, i);
+        accumulator->read(*result, i);
+      }
+    }
+  }
+
+  void extractAccumulators(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    extractValues(groups, numGroups, result);
+  }
+
+  void destroy(folly::Range<char**> groups) override {
```

Comment:
is there anything special about this method? Why isn't this defined in the base class?

Reply:
@pedroerp This method requires knowing the type of the accumulator, which is not available in the base class. Also, some aggregate functions require more logic in 'destroy'.